### DISCUSSION TITLE
is it faster to call find before calling captures?

### DISCUSSION THREAD

**Discussion by tsoutsman** [rust-lang/regex#803] [Q&A] [answered]
Hi,
The documentation and #389 suggest that you should use `find` instead of `captures` as it is more performant. I want to find the line number of a particular regex match and get the capture groups from it. Currently, I'm iterating through the lines of a file and running `captures` on each line, even though most of them won't contain a match. Should I run `find` on each line and then only run `capture` if `find` returns a match?

I've been looking into the `grep` crate which is behind `ripgrep` and that seems to be what it does (at least from the example on [this](https://docs.rs/grep-searcher/0.1.8/grep_searcher/index.html) page).

**Comment by BurntSushi** [comment, marked as answer]
Good question. No, you definitely do not want to run `find` first, as the [regex crate will actually do that for you](https://github.com/rust-lang/regex/blob/d6bc7a4c3b58e1d618024aaededa722df32fa6e8/src/exec.rs#L628-L647). In cases where the DFA can't be used, then you would wind up making your total time twice as slow as you would just be running the NFA twice.

The example in the `grep` crate is doing something different. The searcher only reports _lines_ that match. So you _have_ to run the regex again to find the bounds of each individual match, which is strictly more work. Finding the bounds of the line that matches can be done with less work via the [`shortest_match`](https://docs.rs/regex/1.5.4/regex/struct.Regex.html#method.shortest_match) method. Indeed, that is what the grep crates do internally.

**Reply by tsoutsman** [reply]
Thanks!
